Replicate element coloring with icons

AbsurdlySuspicious opened the pull request:
Fixes #301 
![dawn_icons_preview](https://user-images.githubusercontent.com/46577212/90467416-d3fb0a00-e11c-11ea-911b-7582d2b2e5e7.png)

AbsurdlySuspicious commented on app/src/main/res/values/userpreferences.xml:
```diff
@@ -32,6 +32,9 @@
   <string name="userprefs_show_colored_comments_tree_off">Same color for all lines</string>
 
+  <string name="userprefs_show_color_replication_icons">Color-blind friendly mode</string>
+  <string name="userprefs_show_color_replication_icons_summary">Replicate any significant color with icons</string>
```
Not sure if this is the best way to explain what does it actually do in one sentence

Tunous commented:
Do we actually need a setting for this feature? These icons look very nice and don't seem to clutter the view. It seems to me like they could be always on.

AbsurdlySuspicious commented:
I think we can just rephrase and enable it by default, something like "More icons", because I'm sure some people like cleaner look more.

Tunous commented:
I'm personally more into "Less settings is better" thinking nowadays in aspect of UI and UX. The argument for it is based on the fact that growing amount of possible settings increases code and UX complexity. Not only we have more code paths to maintain, we make it harder (marginally but still) for the app to define its own distinguishing style. As settings like this will grow the app will become more and more confusing for the majority of users entering settings screen.

In this case I don't see huge value in a new setting and don't see how the interface becomes cluttered with new icons, but perhaps I'm missing something 😄

But if others disagree with me then let's go with enabling it by default with changed text.

AbsurdlySuspicious commented:
I removed preference from this branch. When time comes for the next release, if nobody else will voice their opinion, let's merge it as is.

I also added some vertical offset for icons to center them properly. Span centers them between borders of tallest and lowest glyph so with the "common case" text they was visually skewed to the top. This offset should look identical on all devices

binocry commented:
I hope the icons are the same as reddit so users have the same feeling when moving from reddit website to this app

Tunous commented on app/src/main/java/me/saket/dank/widgets/span/CenterAlignedImageSpan.kt:
```diff
@@ -0,0 +1,59 @@
+  }
+
+  // Reimplemented because getCachedDrawable above uses field from this class
+  override fun getSize(paint: Paint?, text: CharSequence?,
```
And if we get rid of caching then we no longer have to override this method.

AbsurdlySuspicious commented on app/src/main/java/me/saket/dank/widgets/span/CenterAlignedImageSpan.kt:
```diff
@@ -0,0 +1,59 @@
+  private var mDrawableRef: WeakReference<Drawable>? = null
+
+  // Reimplemented because this is a private member
+  private fun getCachedDrawable(): Drawable? {
```
I've just pasted it as it was in sdk to avoid any unexpected bugs. Perhaps you're right, I'll look in and fix it in next couple of days

AbsurdlySuspicious commented:
> Span centers them between borders of tallest and lowest glyph

Looks like this approach is inconsistent between different fonts and current implementation is optimal exclusively for Roboto. On the contrary, [`ALIGN_CENTER` implementation from api 29 sdk](https://github.com/AndroidSDKSources/android-sdk-sources-for-api-level-29/blob/master/android/text/style/DynamicDrawableSpan.java#L168) is inconsistent between single and multi-line textviews due to [extra line spacing](https://github.com/Tunous/Dawn/blob/master/app/src/main/res/layout/list_item_submission_content.xml#L37) and I'm not sure which one is better. I also tried icon font approach (using `TypefaceSpan`) back then and it looks terrible. Any ideas?

Tunous commented:
I've pushed a modification where I'm passing in the extra line spacing where it's used. Not the best solution but it appears to look correctly with the fonts I've tested (didn't check all).

AbsurdlySuspicious commented on app/src/main/java/me/saket/dank/widgets/span/CenterAlignedImageSpan.kt:
```diff
@@ -46,14 +30,25 @@ class CenterAlignedImageSpan(d: Drawable): ImageSpan(d, ALIGN_BASELINE) {
-    val b = getCachedDrawable()
+    val drawable = getCachedDrawable()
+    val drawableHeight = drawable.bounds.height()
+    val translationY = (bottom - top - drawableHeight - lineSpacingExtra) / 2f
```
The thing is extra line spacing is not applied to the last line of text. For some reason, it's included in `bottom` value (unlike ordinary line spacing) and thus it's different for last line and all other lines. So this change just "inverts" vertical skew, making icon offset negative for single-line textviews

Tunous replied:
Good to know.

AbsurdlySuspicious replied:
Sorry, I forgot to mark this as resolved, I was still thinking what to do with it at the moment of posting this :)

AbsurdlySuspicious commented on app/src/main/java/me/saket/dank/widgets/span/CenterAlignedImageSpan.kt:
```diff
@@ -13,7 +13,8 @@ class CenterAlignedImageSpan(drawable: Drawable, private val lineSpacingExtra: I
                     top: Int, y: Int, bottom: Int, paint: Paint) {
     val drawableHeight = drawable.bounds.height()
-    val translationY = (bottom - top - drawableHeight - lineSpacingExtra) / 2f
+    val realExtraSpacing = if (canvas.clipBounds.bottom != bottom) lineSpacingExtra else 0 // extra line spacing is not applied to last line
```
dirty but works

AbsurdlySuspicious commented on app/src/main/java/me/saket/dank/widgets/span/CenterAlignedImageSpan.kt:
```diff
@@ -13,7 +13,8 @@ class CenterAlignedImageSpan(drawable: Drawable, private val lineSpacingExtra: I
     val drawableHeight = drawable.bounds.height()
-    val translationY = (bottom - top - drawableHeight - lineSpacingExtra) / 2f
+    val realExtraSpacing = if (canvas.clipBounds.bottom != bottom) lineSpacingExtra else 0 // extra line spacing is not applied to last line
+    val translationY = top + (bottom - top - drawableHeight - realExtraSpacing) / 2f
```
`top +` allows to use this span in lines other than first

Tunous reviewed (approved):
For me looks good

msfjarvis commented on app/src/main/java/me/saket/dank/utils/ColorReplicationIcons.kt:
```diff
@@ -0,0 +1,37 @@
+
+object ColorReplicationIcons {
+  @JvmStatic fun pushIcon(context: Context, builder: Truss, sizeDimenResId: Int,
+                          drawableResId: Int, tintColor: Int, lineSpacingExtra: Int) {
```
not sure what the project policy is but annotating these parameters with `@DrawableRes`, `@ColorRes` and `@DimenRes` would probably not be a bad idea.

Tunous replied:
I'm voting for including these :)

AbsurdlySuspicious replied:
Sorry for such a long delay, I'll address it in a separate PR

msfjarvis reviewed (approved):
I've been using it for a while and haven't found any obvious flaws with either the code or the actual functionality.

elenmerbau commented:
Well, guess it's too late to comment on this. To me these arrows seem like an unnecessary duplication.
